numpy/core/_asarray.pyi

```diff
@@ -0,0 +1,76 @@
+import sys
+from typing import TypeVar, Union, Iterable, overload
+
+from numpy import ndarray, _OrderKACF
+from numpy.typing import ArrayLike, DtypeLike
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+_ArrayType = TypeVar("_ArrayType", bound=ndarray)
+
+def asarray(
+    a: object,
+    dtype: DtypeLike = ...,
+    order: _OrderKACF = ...,
+    *,
+    like: ArrayLike = ...
+) -> ndarray: ...
+@overload
+def asanyarray(
+    a: _ArrayType,
```

Comment:
`def(a: _ArrayType, ...) -> _ArrayType` implies that both the array (sub-)class, dtype and shape are identical for the input and output.

Without higher-kindred typevar we, unfortunately, can not describe a situation were an `ndarray` subclass is returned with a dtype that is different from the initial input (hence the `dtype: None` in this overload).